go/vt/vtadmin/cluster/cluster.go

```diff
@@ -373,50 +374,269 @@ func (c *Cluster) getTablets(ctx context.Context) ([]*vtadminpb.Tablet, error) {
+// (3) Irrespective of whether we're including nonserving shards, or whether
+// we're doing size aggregation at all, we will only make GetSchema RPCs to
+// tablets that are in SERVING state; we don't want to use a tablet that might
+// be in a bad state as the source of truth for a schema. Therefore if we can't
+// find a SERVING tablet for the keyspace (in non-aggregation mode) or for a
+// shard in that keyspace (in aggregation mode), then we will return an error
+// back to the caller.
+func (c *Cluster) GetSchema(ctx context.Context, keyspace string, opts GetSchemaOptions) (*vtadminpb.Schema, error) {
 	span, ctx := trace.NewSpan(ctx, "Cluster.GetSchema")
 	defer span.Finish()
 
+	if opts.SizeOpts == nil {
+		opts.SizeOpts = &vtadminpb.GetSchemaTableSizeOptions{
+			AggregateSizes:          false,
+			IncludeNonServingShards: false,
+		}
+	}
+
+	if opts.BaseRequest == nil {
+		opts.BaseRequest = &vtctldatapb.GetSchemaRequest{}
+	}
+
+	if opts.SizeOpts.AggregateSizes && opts.BaseRequest.TableNamesOnly {
+		log.Warningf("GetSchema(cluster = %s) size aggregation is incompatible with TableNamesOnly, ignoring the latter in favor of aggregating sizes", c.ID)
+		opts.BaseRequest.TableNamesOnly = false
+	}
+
 	AnnotateSpan(c, span)
+	span.Annotate("keyspace", keyspace)
+	annotateGetSchemaRequest(opts.BaseRequest, span)
+	vtadminproto.AnnotateSpanWithGetSchemaTableSizeOptions(opts.SizeOpts, span)
 
-	// Copy the request to not mutate the caller's request object.
-	r := *req
-	r.TabletAlias = tablet.Tablet.Alias
+	if len(opts.Tablets) == 0 {
+		// Fetch all tablets for the keyspace.
+		var err error
 
-	span.Annotate("tablet_alias", topoproto.TabletAliasString(r.TabletAlias))
-	span.Annotate("exclude_tables", strings.Join(r.ExcludeTables, ","))
-	span.Annotate("tables", strings.Join(r.Tables, ","))
-	span.Annotate("include_views", r.IncludeViews)
-	span.Annotate("table_names_only", r.TableNamesOnly)
-	span.Annotate("table_sizes_only", r.TableSizesOnly)
+		opts.Tablets, err = c.FindTablets(ctx, func(tablet *vtadminpb.Tablet) bool {
+			return tablet.Tablet.Keyspace == keyspace
+		}, -1)
+		if err != nil {
+			return nil, fmt.Errorf("%w for keyspace %s", errors.ErrNoTablet, keyspace)
+		}
+	}
 
-	schema, err := c.Vtctld.GetSchema(ctx, &r)
-	if err != nil {
-		return nil, err
+	if err := c.Vtctld.Dial(ctx); err != nil {
+		return nil, fmt.Errorf("failed to Dial vtctld for cluster = %s for GetSchema: %w", c.ID, err)
 	}
 
-	if schema == nil || schema.Schema == nil {
-		return nil, nil
+	var tabletsToQuery []*vtadminpb.Tablet
+
+	if opts.SizeOpts.AggregateSizes {
+		span, ctx := trace.NewSpan(ctx, "Cluster.FindAllShardsInKeyspace")
+
+		AnnotateSpan(c, span)
+		span.Annotate("keyspace", keyspace)
+
+		resp, err := c.Vtctld.FindAllShardsInKeyspace(ctx, &vtctldatapb.FindAllShardsInKeyspaceRequest{
+			Keyspace: keyspace,
+		})
+
+		span.Finish()
```

Comment:
Really minor + non-blocking point -- did you consider a `(c *Cluster) FindAllShardsInKeyspace` function? That might simplify this a tiny bit, especially with the annotations. (It took me 0.2 seconds to realize that in this case we _don't_ want a `defer span.Finish()` call.) 😊

Reply:
yeah, that sounds good! (i didn't because i didn't want to write more tests :shamebell:)